Upgrade i18next to latest

40thieves opened the pull request:
### Description

Upgrades i18next to v19, and change the config/set up to match the new API.

### Related Issues / PRs

Closes https://github.com/overleaf/issues/issues/3276

### Review

Probably best reviewed commit-by-commit.

I don't believe there are any changes needed on the web side. I've tried to keep this backwards compatible.

### Manual Testing

Installed the branch to a local version of web and checked

- [x] English translations work
- [x] French translations work
- [x] Interpolation works
  - [x] Basic
  - [x] With multiple interpolations (e.g. `Hello __foo__-world __bar__`)
  - [x] With escaping disabled
- [x] `showUserOtherLng` flag works

das7pad commented on app/js/translations.js:
```diff
@@ -10,14 +10,18 @@ module.exports = {
+      // underscores instead of curly braces
+      interpolation: {
+        prefix: '__',
+        suffix: '__'
```
Note that this will break translations from locales with a `-` next to `__`. IIRC there is yet another option for changing the escape character or something like that.

`pre __appName__- middle __otherField__ post` is omitting the `middle` part.

40thieves replied:
Thanks, it was on my list to check exactly what the interpolation settings were for the old syntax. The docs point here: https://github.com/i18next/i18next/blob/5fb4dbcb7ccdcfc78192e69ef3291d2ace6814b4/test/backward/compatibility/v1.js#L4 for how they convert their old tests.

40thieves commented on app/js/translations.js:
```diff
@@ -35,24 +43,31 @@ module.exports = {
+    const expressMiddleware = middleware.handle(i18n)
+
+    function backwardsCompatExpressMiddleware(req, res, next) {
+      req.i18n.translate = req.i18n.t
```
I'm unsure if this is a good idea or not, but would greatly reduce the diff on the web side. Will revisit on Monday.

das7pad replied:
Let's move this into line 56 -- that should require no changes in web (other than the package-lock)?

40thieves replied:
I did try this on Friday, but was too distracted by the heat to get it working 😅  It's now done in https://github.com/overleaf/translations/pull/31/commits/3ec5127fd4e2d98dfb007f9acbd4e627d5e88e98.

40thieves commented:
Yes I just pushed a commit to match that, this PR is still WIP.

I was leaning towards a change in `ExpressLocals` to disable where needed (hence separate commits). But then I realised there's also direct calls to `req.i18n.translate`, where we might be relying on the lack of escaping. So I've switched back to disabling it by default.

40thieves commented:
Fwiw, I also decided to leave off https://github.com/i18next/i18next/blob/5fb4dbcb7ccdcfc78192e69ef3291d2ace6814b4/test/backward/compatibility/v1.js#L13-L14 since I can't find any usage of the nesting syntax.

das7pad commented:
We probably want to disable the nested interpolation as well? It can be used to blow-up translations/mislead users quite a bit: https://repl.it/repls/SimilarImpressiveScriptinglanguages (forked from your repl, nice tool for this kind of thing!)

https://www.i18next.com/translation-function/nesting Oh and it supports passing your own context in -- hello prototype pollution.

40thieves commented:
> We probably want to disable the nested interpolation as well?

To be clear - I meant that I couldn't find any usage of nesting in our existing translations, and so we don't need to add backwards compatibility for it.

However as you point out there is a security angle here. I've added the `skipOnVariables` flag which prevents nesting inside interpolated values.

40thieves commented:
I think this is now ready for review. Assigning @mserranom for fresh eyes and @das7pad since you've already looked at it.

40thieves commented:
> i18n.translate is not exported

Ah right, I hadn't realised that it was used directly. I'll fix that now.

40thieves commented:
So I looked into the failing test (`should not try to extract the project name from the URI`), with the fix for `i18n.translate` and it's still failing (locally and on CI).

I believe it may actually be a problem with the test, and the old behaviour is only "accidentally" working. This is what I think is happening:

1. The test bootstrap initialises the server: https://github.com/overleaf/web-internal/blob/35c87efcbc1add4d4731c371262c69f5ab646b9a/test/acceptance/src/Init.js#L1
2. This ends up running the translation package's `setup` with some settings: https://github.com/overleaf/web-internal/blob/35c87efcbc1add4d4731c371262c69f5ab646b9a/app/src/infrastructure/Server.js#L31
3. The test then runs `setup` again, this time with no settings: https://github.com/overleaf/web-internal/blob/35c87efcbc1add4d4731c371262c69f5ab646b9a/modules/open-in-overleaf/test/acceptance/src/OpenInOverleafTests.js#L18

With the old version, I suspect that the translations are cached so that on the second call the lack of settings is ignored, and the English translations are actually loaded within the test.

In the latest version, no translations are found within the test (due to the lack of settings). This means that `translate` will only echo back the key. The test relies on the English translation being loaded (and not being changed), so fails.

There are a few options here:

1.  Switch the assertion to `translations.i18n.translate('new_snippet_project')`. This is what we are doing in at least 1 other test: https://github.com/overleaf/web-internal/blob/35c87efcbc1add4d4731c371262c69f5ab646b9a/modules/open-in-overleaf/test/acceptance/src/OpenInOverleafTests.js#L1539
2. Pass some minimal config to the `setup` call within the test, so that the English translations are loaded correctly
3. Create some mechanism for accessing the `i18n` instance created in the server initialisation

I am currently leaning towards option 1, as this would protect the test against any changes to the underlying translation string. However in the long term, I think we should also consider option 2 or 3.

cc @gh2k @jdleesmiller

40thieves commented:
I went ahead a made a PR with the test fix proposed above: https://github.com/overleaf/web-internal/pull/3076. This is blocked until that is resolved.

jdleesmiller commented:
OK, option 1 seems fine for now. I would lean toward option 2 in the longer term --- having actual (English) strings hopefully makes the tests a bit easier to read and write and brings them closer to what the user actually sees (even though that does mean that changing translations can break tests).

40thieves commented:
I was thinking more about this and I realised that there's an option 4: remove the translations package entirely and test only against the English strings.

This has a few benefits:

1. Satisfy the desire to make the tests closer to what the user sees
2. Bring the tests more inline with the other tests (AFAICT this is the only instance of the package being imported to the tests)
3. We could remove the `i18n.translate` alias that was added only for this one test

This is another longer term fix as it would require changing the existing tests to use the English strings.

40thieves commented:
> The setLng logic seems to be broken

Thanks good catch. Fixed in https://github.com/overleaf/translations/pull/31/commits/4b958d0d04c2afc6797bc864d77b70357da033e8, and added test for this behaviour.

das7pad commented on app/js/translations.js:
```diff
@@ -1,28 +1,61 @@
     function setLangBasedOnDomainMiddleware(req, res, next) {
+      // setLng query param takes precedence, so if set ignore the subdomain
+      if (req.originalUrl.includes('setLng')) {
+        return next()
```
Sorry to bother on `setLng` again. `setLng` was previously triggering the `showUserOtherLng` logic as well. Shall we restore this? (Line 36-37)

40thieves replied:
Hmm, right.

The only place I could find usage of the `setLng` query param was in the smoke tests: https://github.com/overleaf/web-internal/blob/8335d3185075ed4049931c29c37cfa8bbb42ee53/test/smoke/src/SmokeTests.js#L31. I'm not aware of any users relying on this and I failed to find any hits searching the ingress logs for the last few days (stackdriver keeps erroring beyond that).

So I think I would lean towards removing this?

40thieves replied:
My mental model is that `setLng` is an override, so we should just set the language and move on.

das7pad replied:
> So I think I would lean towards removing this?

That's fine by me.

40thieves replied:
Thanks. Could you approve the PR?

40thieves commented:
Oh actually, before you approve, I have some tests that I wanted to add.

das7pad commented:
> Oh actually, before you approve, I have some tests that I wanted to add.

Snap I already submitted, but you can probably add them w/o my approval :) (Or just ask for another review, I do not mind.)

40thieves commented:
Ok, I've added some tests for the translate function, which I believe addresses most of https://github.com/overleaf/issues/issues/3390. I talked it through with @jdleesmiller and we think it's fine to just test for the actual translation string (even though there's some risk of the translation string changing, causing the test to fail).
